engine/resource/src/resource_mounts.cpp

```diff
@@ -0,0 +1,539 @@
+    {
+        dmLogError("Loader can't mount uri '%s'", path);
+        return dmResource::RESULT_NOT_SUPPORTED;
+    }
+
+    dmResourceProvider::HArchive archive;
+    dmResourceProvider::Result provider_result = dmResourceProvider::CreateMount(loader, &uri, ctx->m_ResourceBaseArchive, &archive);
+    if (dmResourceProvider::RESULT_OK != provider_result)
+    {
+        dmLogError("Failed to create mount '%s' - '%s': %d", name, path, provider_result);
+        return dmResource::RESULT_IO_ERROR;
+    }
+
+    dmResourceMounts::AddMount(ctx, name, archive, priority, persist);
+
+    return dmResource::RESULT_OK;
+}
+
+dmResource::Result LoadMounts(HContext ctx, const char* app_support_path)
+{
+    char path[DMPATH_MAX_PATH];
+    dmPath::Concat(app_support_path, MOUNTS_FILENAME, path, sizeof(path));
+
+    if (!dmSys::Exists(path))
+    {
+        DM_RESOURCE_DBG_LOG(1, "No mount file detected: '%s'\n", path);
+        return dmResource::RESULT_OK;
+    }
+
+    DM_RESOURCE_DBG_LOG(1, "Loading mounts '%s'\n", path);
+
+    DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex); // avoid reading while it might be written
+
+    dmArray<MountFileEntry> entries;
+    dmResource::Result result = ReadMountsFile(path, entries);
+    if (dmResource::RESULT_OK != result)
+    {
+        dmLogError("Failed to read mounts file");
+        return result;
+    }
+
+    uint32_t size = entries.Size();
+    for (uint32_t i = 0; i < size; ++i)
+    {
+        MountFileEntry& entry = entries[i];
+
+        DM_RESOURCE_DBG_LOG(2, "  mounting: '%s' %d '%s'\n", entry.m_Name, entry.m_Priority, entry.m_Uri);
+        LoadMount(ctx, entry.m_Priority, entry.m_Name, entry.m_Uri, true);
+    }
+
+    FreeMountsFile(entries);
+    return dmResource::RESULT_OK;
+}
+
+dmResource::Result SaveMounts(HContext ctx, const char* app_support_path)
+{
+    char path[DMPATH_MAX_PATH];
+    dmPath::Concat(app_support_path, MOUNTS_FILENAME, path, sizeof(path));
+
+    DM_RESOURCE_DBG_LOG(1, "Saving mounts '%s'\n", path);
+
+    DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex);
+
+    dmArray<MountFileEntry> entries;
+
+    uint32_t size = ctx->m_Mounts.Size();
+    for (uint32_t i = 0; i < size; ++i)
+    {
+        ArchiveMount& mount = ctx->m_Mounts[i];
+        if (!mount.m_Persist)
+            continue;
+
+        if (entries.Full())
+            entries.OffsetCapacity(8);
+
+        dmURI::Parts uri;
+        dmResourceProvider::GetUri(mount.m_Archive, &uri);
+
+        char uri_str[1024];
+        if (uri.m_Location[0] == '\0')
+            dmSnPrintf(uri_str, sizeof(uri_str), "%s:%s", uri.m_Scheme, uri.m_Path);
+        else
+            dmSnPrintf(uri_str, sizeof(uri_str), "%s:%s/%s", uri.m_Scheme, uri.m_Location, uri.m_Path);
+
+        MountFileEntry entry;
+        entry.m_Name = strdup(mount.m_Name);
+        entry.m_Uri = strdup(uri_str);
+        entry.m_Priority = mount.m_Priority;
+
+        entries.Push(entry);
+    }
+
+    dmResource::Result result = WriteMountsFile(path, entries);
```

Comment:
I split the load/save into a separate file, to make it less cluttered here, and also more unit testable.